### PR TITLE
Support multiple pvt regions for drsdtcon

### PR DESCRIPTION
**Pull request by totto82**
Most importantly drsdtcon can now be deactivated locally by setting first item to < 0

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this opm-simulators=5527 please

**Comment by totto82** [comment]
jenkins build this opm-simulators=5527 please

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 please

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 please

**Comment by GitPaean** [comment]
I am not familiar with multiple PVT regions with SCHEDULE keywords. 

For PROPS keyword (like PVTO), if there are multiple PVT regions, if the DATA for the first region is input and the later ones are defaulted, other regions will use the information from the prior regions.

In the current implementation, DRSDTCON will use default values if later regions does not specify. 

What will be the desirable behavoir here to you?

**Comment by GitPaean** [comment]
The PR touched briefly implementation related to DRSDT and DRVDT. Do those keywords also support multiple PVT regions now?

File: tests/parser/ScheduleTests.cpp [hunk 1]
```diff
@@ -2043,20 +2052,20 @@ VAPPARS
 
     const OilVaporizationProperties& ovap1 = schedule[1].oilvap();
     BOOST_CHECK(ovap1.getType() == OilVaporizationProperties::OilVaporization::DRDT);
-    BOOST_CHECK_EQUAL(true,   ovap1.drsdtActive());
```
Comment (GitPaean) [inline tests/parser/ScheduleTests.cpp:2046]: The functions without parameters are still there. Then can still be tested.

**Comment by totto82** [comment]
> The PR touches briefly implementation related to DRSDT and DRVDT. Do those keywords also support multiple PVT regions now?

Yes by using DRSDTR and DRVDTR.

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 please

**Comment by totto82** [comment]
@GitPaean Thanks for the review. I have addressed your comments here and in simulators.

**Comment by GitPaean** [comment]
update_data most likely will not work, while I would like to see the regression failures.

**Comment by GitPaean** [comment]
jenkins build this opm-models=924 opm-simulators=5527 opm-tests=1209 update_data please

**Comment by GitPaean** [comment]
> > The PR touches briefly implementation related to DRSDT and DRVDT. Do those keywords also support multiple PVT regions now?
> 
> Yes by using DRSDTR and DRVDTR.

```c++
void handleDRVDT(HandlerContext& handlerContext)
{
    const std::size_t numPvtRegions = handlerContext.static_schedule().m_runspec.tabdims().getNumPVTTables();
    std::vector<double> maximums(numPvtRegions);
    for (const auto& record : handlerContext.keyword) {
        const auto& max = record.getItem<ParserKeywords::DRVDTR::DRVDT_MAX>().getSIDouble(0);
        std::fill(maximums.begin(), maximums.end(), max);
        auto& ovp = handlerContext.state().oilvap();
        OilVaporizationProperties::updateDRVDT(ovp, maximums);
    }
}```

The parsing of `DRVDT` does not look like handling the PVT regions correctly?

**Comment by GitPaean** [comment]
> > > The PR touches briefly implementation related to DRSDT and DRVDT. Do those keywords also support multiple PVT regions now?
> > 
> > 
> > Yes by using DRSDTR and DRVDTR.
> 
> ```c++
> void handleDRVDT(HandlerContext& handlerContext)
> {
>     const std::size_t numPvtRegions = handlerContext.static_schedule().m_runspec.tabdims().getNumPVTTables();
>     std::vector<double> maximums(numPvtRegions);
>     for (const auto& record : handlerContext.keyword) {
>         const auto& max = record.getItem<ParserKeywords::DRVDTR::DRVDT_MAX>().getSIDouble(0);
>         std::fill(maximums.begin(), maximums.end(), max);
>         auto& ovp = handlerContext.state().oilvap();
>         OilVaporizationProperties::updateDRVDT(ovp, maximums);
>     }
> }
> ```
> 
> The parsing of `DRVDT` does not look like handling the PVT regions correctly?

sorry. I see DRVDTR and DRVDT are different keywords now.

**Comment by GitPaean** [comment]
> @GitPaean Thanks for the review. I have addressed your comments here and in simulators.

Thanks for the update. I am happy with the PR now.

**Comment by totto82** [comment]
Thanks for approving. I can do the merge. I will start with the code and then do the test.

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 please

**Comment by GitPaean** [comment]
> Thanks for approving. I can do the merge. I will start with the code and then do the test.

That is okay. 

Since we are modifying the existing DATA file in opm-tests, make sure the modification and the update reference will go in together.   Although personally, I think it is nicer that the simulator code modification and the update reference can go in together, otherwise, we basically not testing the simulator code modification. 

Another point, make sure the manual for the keyword DRSDTCON will be correct when coming to the default behavoir when coming to multiple PVT region setup.  It is different with the multiple PVT region PROPS keywords (it is stated clearly in the manual for multiple PVT region PROPS keyword.)

**Comment by totto82** [comment]
This needs a rebase after the moving of mixControl_ I will update asap.

**Comment by GitPaean** [comment]
> This needs a rebase after the moving of mixControl_ I will update asap.

sorry for the trouble. Did not think about it.  we probably need to rebase one or the other.

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 please

**Comment by akva2** [comment]
mind the warnings.

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 please

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 please

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 opm-tests=1209 please

**Comment by bska** [comment]
jenkins build this opm-models=924 opm-simulators=5527 opm-tests=1209 please

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 opm-tests=1209 update_data please

**Comment by bska** [comment]
jenkins build this opm-models=924 opm-simulators=5527 opm-tests=1209 update_data please

**Comment by totto82** [comment]
Have I lost the right to start jenkins? Why can't I trigger it.

**Comment by akva2** [comment]
no idea, you have not been removed from list and there is nothing in the logs.

**Comment by GitPaean** [comment]
In my experience, it happens from time to time. You just need to try again.

**Comment by akva2** [comment]
this won't work. the PR in opm-tests you use already modify files that the update_data wants to modify. that is not a supported scenario.

**Comment by GitPaean** [comment]
so if the update_data commit in https://github.com/OPM/opm-tests/pull/1209 be removed, it should work right?

I think we got the update_data commit that way. Somehow the results changed again maybe due to update of the master branch.

**Comment by akva2** [comment]
yes then it should be fine.

**Comment by totto82** [comment]
> this won't work. the PR in opm-tests you use already modify files that the update_data wants to modify. that is not a supported scenario. 

Cant I cherry-pick it and merge it with the earlier update and then push to opm-tests1209

**Comment by akva2** [comment]
cherry-pick what exactly? it failed to generate a proper data update because there is already an update for those files in opm tests 1209.

**Comment by totto82** [comment]
I noticed. I have removed the update commit from the test. Lets try again.

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 opm-tests=1209 update_data please

**Comment by totto82** [comment]
jenkins build this opm-models=924 opm-simulators=5527 opm-tests=1209 please

**Comment by totto82** [comment]
The PRs are approved and Jenkins is happy. I will start the merging process.

**Pull request merged**